homeassistant/components/volvooncall.py

```diff
@@ -81,56 +115,76 @@ def setup(hass, config):
 
-    state = hass.data[DATA_KEY] = VolvoData(config)
+    data = hass.data[DATA_KEY] = VolvoData(config)
 
     def discover_vehicle(vehicle):
         """Load relevant platforms."""
-        state.entities[vehicle.vin] = []
-        for attr, (component, *_) in RESOURCES.items():
-            if (getattr(vehicle, attr + '_supported', True) and
-                    attr in config[DOMAIN].get(CONF_RESOURCES, [attr])):
-                discovery.load_platform(
-                    hass, component, DOMAIN, (vehicle.vin, attr), config)
-
-    def update_vehicle(vehicle):
-        """Receive updated information on vehicle."""
-        state.vehicles[vehicle.vin] = vehicle
-        if vehicle.vin not in state.entities:
-            discover_vehicle(vehicle)
-
-        for entity in state.entities[vehicle.vin]:
-            entity.schedule_update_ha_state()
-
-        dispatcher_send(hass, SIGNAL_VEHICLE_SEEN, vehicle)
-
-    def update(now):
+        data.vehicles.append(vehicle.vin)
+
+        dashboard = vehicle.dashboard(
+            mutable=config[DOMAIN][CONF_MUTABLE],
+            scandinavian_miles=config[DOMAIN][CONF_SCANDINAVIAN_MILES])
+
+        def is_enabled(attr):
+            return attr in config[DOMAIN].get(CONF_RESOURCES, [attr])
+
+        for instrument in (
+                instrument
+                for instrument in dashboard.instruments
+                if instrument.component in COMPONENTS and
+                is_enabled(instrument.slug_attr)):
+
+            data.instruments.append(instrument)
+
+            hass.async_create_task(
+                discovery.async_load_platform(
+                    hass,
+                    COMPONENTS[instrument.component],
+                    DOMAIN,
+                    (vehicle.vin,
+                     instrument.component,
+                     instrument.attr),
+                    config))
+
+    async def update(now):
         """Update status from the online service."""
         try:
-            if not connection.update():
+            if not await connection.update(journal=True):
                 _LOGGER.warning("Could not query server")
                 return False
 
             for vehicle in connection.vehicles:
-                update_vehicle(vehicle)
+                if vehicle.vin not in data.vehicles:
+                    discover_vehicle(vehicle)
+
+            async_dispatcher_send(hass, SIGNAL_STATE_UPDATED)
 
             return True
         finally:
-            track_point_in_utc_time(hass, update, utcnow() + interval)
+            async_track_point_in_utc_time(hass, update, utcnow() + interval)
 
     _LOGGER.info("Logging in to service")
-    return update(utcnow())
+    return await update(utcnow())
 
 
 class VolvoData:
     """Hold component state."""
 
     def __init__(self, config):
         """Initialize the component state."""
-        self.entities = {}
-        self.vehicles = {}
+        self.vehicles = []
```

Comment:
If this should only contain unique vins, we could make it a `set` to speed up lookup and membership check.
https://docs.python.org/3/library/stdtypes.html#types-set
```suggestion
        self.vehicles = set()
```

Reply:
Of course, we have to consider the poor user with 4711 Volvo cars in the garage. :-) We cannot have average O(n) complexity for lookups here!
(fixed)

Reply:
Removing WIP (but still missing documentation)